src/sctools/fastq.py

```diff
@@ -0,0 +1,197 @@
+
+    Defines properties for accessing fastq record information:
+    :property name: name field
+    :property sequence: sequence field
+    :property name2: second name field
+    :property quality: quality field
+
+    Utilities for parsing read attributes:
+    :property average_quality: return the mean quality of FastqRecord
+    """
+    __slots__ = ['_data']
+
+    def __init__(self, record):
+        """
+        :param [str|bytes] record: list of four strings or bytes objects defining a
+          single fastq record
+        """
+        self._data = list(record)
+
+    @property
+    def name(self):
+        return self._data[0]
+
+    @name.setter
+    def name(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('name must be str or bytes')
+        else:
+            self._data[0] = value
+
+    @property
+    def sequence(self):
+        return self._data[1]
+
+    @sequence.setter
+    def sequence(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('sequence must be str or bytes')
+        else:
+            self._data[1] = value
+
+    @property
+    def name2(self):
+        return self._data[2]
+
+    @name2.setter
+    def name2(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('name2 must be str or bytes')
+        else:
+            self._data[2] = value
+
+    @property
+    def quality(self):
+        return self._data[3]
+
+    @quality.setter
+    def quality(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('quality must be str or bytes')
+        else:
+            self._data[3] = value
+
+    def __bytes__(self):
+        try:
+            return b''.join(self._data)
+        except TypeError:
+            return ''.join(self._data).encode()
+
+    def __str__(self):
+        try:
+            return ''.join(self._data)
+        except TypeError:
+            return b''.join(self._data).decode()
+
+    def __repr__(self):
+        return "Name: %s\nSequence: %s\nName2: %s\nQuality: %s\n" % (
+            self._data[0], self._data[1], self._data[2], self._data[3])
+
+    def __len__(self):
+        return len(self.sequence)
+
+    def average_quality(self):
+        raise NotImplementedError
+
+
+class BytesRecord(Record):
+    """Fastq record object
+
+    Defines several properties for accessing fastq record information:
+    :property bytes name: name field
+    :property bytes sequence: sequence field
+    :property bytes name2: second name field
+    :property bytes quality: quality field
+
+    Also defines several methods for accessing SEQC annotation fields:
+    :property float average_quality: return the mean quality of FastqRecord
+    """
+
+    def average_quality(self):
+        """return the average quality of this record"""
+        return sum(c for c in self.quality[:-1]) / (len(self.quality[:-1]) - 1) - 33
```

Comment:
Why minus 33? A comment might be helpful here.

Reply:
Absolutely correct. It's because of an old solexa/illumina conversion.